### PR TITLE
Open a specific wiki when passed on the command line

### PR DESCRIPTION
**Pull request by mklauber**
Fix for #85.  If a file or folder path is specified after ./nw that wiki is used in place of `backstage://WikiListWindow` as the "main" window.  

~~Note: I cannot figure out what the empty value of `gui.App.argv` is, so I am using `!=` rather than `!==`.  Hopefully someone else knows what that might be, and I can correct things.~~

### CODE REVIEW

**Comment by Jermolene** [comment]
Thanks @mklauber, this looks good.

My sincere apologies that your PRs have languished for so long. I've picked up steam on TiddlyDesktop recently and would like to get them merged, and welcome any further enhancements you have in mind.

**Comment by Jermolene** [comment]
Hi @mklauber I just came back to this but notice that in fact the PR is incomplete; you are assigning to the new variable "dest" but not actually using it.

**Comment by JohnHind** [comment]
Hi @Jermolene, any chance you could get this one over the line even if @mklauber has understandably lost interest? Looking back, I first suggested this in February 2014! I tried figuring it out myself, but do not have time to learn the codebase and the @mklauber patches seem to be against an earlier version. Willing to sponsor if it would help, or donate to a nominated charity if preferred (yes, I want this badly enough to resort to bribery!).

File: source/js/main.js [hunk 1]
```diff
@@ -124,9 +124,17 @@ var wikilistWindow;
 $tw.boot.suppressBoot = true;
 require("../tiddlywiki/boot/boot.js").TiddlyWiki($tw);
 $tw.boot.boot(function() {
-	wikilistWindow = $tw.desktop.windowList.openByUrl("backstage://WikiListWindow",{mustQuitOnClose: true});
+  var dest;
+  if (gui.App.argv.length !== 0){
+    if($tw.utils.isDirectory(gui.App.argv)){
```
Comment (BurningTreeC) [inline source/js/main.js]: the argument is missing here ... gui.App.argv[0]

Comment (BurningTreeC) [inline source/js/main.js]: the argument is missing here ... gui.App.argv[0]

**Comment by mklauber** [comment]
Thanks, fixed

**Comment by BurningTreeC** [comment]
Thanks, sorry for my reviewing mess :grin:

**Comment by BurningTreeC** [comment]
Hi @mklauber , opening wikies directly from the command line works fine now, thanks for your work :)

Do you know a way to open more than one wiki at a time with this?

**Comment by JohnHind** [comment]
Thanks @mklauber and @BurningTreeC! Hopefully @Jermolene will now release this in the next version. Meantime, I've patched mine manually and it works a treat.

**Comment by JohnHind** [comment]
Ah - wait, spoke too soon, there is a problem! If there is no file but there is a command line switch, it tries to open that as a file and presents an empty browser window which for some odd reason cannot be closed! Need to investigate further, but I think you need to either check the target is a file, or check it does not start with a command line switch (--). I tried $tw.utils.isFile() by extension, but this did not work and led me back to the old codebase learning problem since I cannot see where the hell $tw.utils.isDirectory is defined! I really hate JavaScript!!

**Comment by Jermolene** [comment]
Hi @JohnHind thanks for your help with this

>  I tried $tw.utils.isFile() by extension, but this did not work and led me back to the old codebase learning problem since I cannot see where the hell $tw.utils.isDirectory is defined! I really hate JavaScript!!

`$tw.desktop.*` is defined in TiddlyDesktop; the main `$tw.*` global is defined in TiddlyWiki5, in this case here:

https://github.com/Jermolene/TiddlyWiki5/blob/ffc0899f52b45ba2b5a5a20fb909dcc587e80b3e/core/modules/utils/filesystem.js#L142

**Comment by JohnHind** [comment]
@BurningTreeC, no that will not work because dest is not defined when you have a command line switch and no filename. This ought to be OK:
```
$tw.boot.boot(function() {
	var dest = "backstage://WikiListWindow";
	var srcPath = gui.App.argv[0];
	var srcStats = fs.lstatSync(srcPath);
	if (gui.App.argv.length !== 0){
	  if($tw.utils.isDirectory(gui.App.argv[0])){
		dest = "wikifolder://" + gui.App.argv[0];
	  } else if(srcStats.isFile()) {
		dest = "wikifile://" + gui.App.argv[0];
	  }
	}
	wikilistWindow = $tw.desktop.windowList.openByUrl(dest,{mustQuitOnClose: true});
});

```
But this crashes too and I do not see why. Do you guys have a way of debugging? When I get a crash I just get no UI and a bunch of nwjs processes that have to be killed in Task Manager!

**Comment by mklauber** [comment]
@JohnHind I think there might be an issue in the `if isFile` line, where you're not passing gui.App.argv[0] to `isFile`.  But I'm not sure how that'd be causing the issue you're describing.

**Comment by mklauber** [comment]
Regardless, I think we may need to rethink our approach to command line arguments here, it feels like we're bandaging over the fact that we need proper parameter parsing.

**Comment by JohnHind** [comment]
This is closer - works with a filename and with no parameter, still crashes with a switch. We need a way of testing if the parameter is a properly formed path before feeding it to fs.lstatSync. Parsing for a switch will not help because we cannot crash for any random input.

```
$tw.boot.boot(function() {
	var dest = "backstage://WikiListWindow";
	var srcStats;
	if (gui.App.argv.length > 0) {
		srcStats = fs.lstatSync(gui.App.argv[0])
		if (srcStats.isFile()){
			dest = "wikifile://" + gui.App.argv[0];
		} else if (srcStats.isDirectory()){
			dest = "wikifolder://" + gui.App.argv[0];
		}
	}
	wikilistWindow = $tw.desktop.windowList.openByUrl(dest,{mustQuitOnClose: true});
  });

})();

```

**Comment by BurningTreeC** [comment]
@JohnHind - I did a grep search for isFile and found that it's used within TiddlyWiki

I just copied this check from TiddlyWiki5/core/modules/utils/filesystem.js - [see](https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/utils/filesystem.js)

**Comment by BurningTreeC** [comment]
@JohnHind - fs and path is already required on top of main.js - so it was convenient to use this check here, too

**Comment by JohnHind** [comment]
OK, some documentation here:
https://nodejs.org/api/fs.html
But the asynchronous architecture makes it incredibly obscurantist!
Over to you guys - it is past my bedtime and my head hurts!

**Comment by BurningTreeC** [comment]
@mklauber 

[srcStats.isFile()](https://nodejs.org/api/fs.html#fs_stats_isfile) is correct in this case, returns true if it's a regular file. The argument is within srcStats through 

srcStats = fs.lstatSync(gui.App.argv[0])

@JohnHind - thanks for the link to the docs, these are very helpful
handling the asynchronous methods is mentioned in the docs [under "File System"](https://nodejs.org/api/fs.html#fs_file_system)

**Comment by JohnHind** [comment]
OK guys, after a good night's sleep I think I cracked it:
```
$tw.boot.boot(function() {
  var dest = "backstage://WikiListWindow";
  var srcStats;
  if (gui.App.argv.length > 0) {
    try {
      srcStats = fs.lstatSync(gui.App.argv[0]);
    } catch(err) {}
    if (srcStats) {
      if (srcStats.isDirectory()) {
        dest = "wikifolder://" + gui.App.argv[0];
      } else if (srcStats.isFile()) {
	dest = "wikifile://" + gui.App.argv[0];
      }
    }
  }
  wikilistWindow = $tw.desktop.windowList.openByUrl(dest,{mustQuitOnClose: true});
});

```
This is good in all my test cases: no parameter, a switch, a filepath. Please test in your own cases, including a directory because I am not sure what is expected in that case.

**Comment by BurningTreeC** [comment]
@JohnHind , this works!

The folder is for wikies running under node.js
One can initialize a wiki folder:

./tiddlywiki.js /path/to/folder --init server

and then serve it to localhost:

./tiddlywiki.js /path/to/folder --server 8080

If passing a directory path to nw now it serves that folder as a wiki folder... renders all files as a tiddlywiki ... if there's a correct tiddlywiki.info file inside that folder. If there's no tiddlywiki.info it serves the folder as it appears.

I think this is all fine

**Comment by mklauber** [comment]
@JohnHind does that work if you do a `--flag filepath`? I.e. when the filepath is the not the first argument?  Because I don't think it does, and I think we should rework this to allow/handle command line arguments in any order.

**Comment by JohnHind** [comment]
@mklauber no, it does not. But I do not think it is unreasonable to require the filepath to be the first argument with any switches following, so long as the documentation makes it clear. Many command lines make the order of parameters significant. In this case there is only one documented command line switch so it is not exactly rocket surgery!

**Comment by mklauber** [comment]
@Jermolene Does the above comment meet your approval?  If so, I'll make the suggested changes and add documentation to the readme file about the importance of ordering in this instance.

**Comment by BurningTreeC** [comment]
Hi all,

if I close a wiki opened with the command line, the process has to be killed (Ctrl+c) - setting mustQuitOnClose: false closes it when I close the wiki

Is it right, to set a variable on top  (mustQuitOnClose: myVar) and set it to false in if(isFile) and if(isDirectory) ?

**Comment by BurningTreeC** [comment]
That's what I mean:

```
$tw.boot.boot(function() {
  var dest = "backstage://WikiListWindow";
  var srcStats;
  var quitOnClose = true;
  if (gui.App.argv.length > 0) {
    try {
      srcStats = fs.lstatSync(gui.App.argv[0]);
    } catch(err) {}
    if (srcStats) {
      if (srcStats.isDirectory()) {
        dest = "wikifolder://" + gui.App.argv[0];
        quitOnClose = false;
      } else if (srcStats.isFile()) {
	dest = "wikifile://" + gui.App.argv[0];
        quitOnClose = false;
      }
    }
  }
  wikilistWindow = $tw.desktop.windowList.openByUrl(dest,{mustQuitOnClose: quitOnClose});
});
```

**Comment by JohnHind** [comment]
@BurningTreeC : What OS are you using? I'm Windows 10 and the process is terminating fine when I close the wiki window (with the code from my post above).

**Comment by BurningTreeC** [comment]
@JohnHind I'm on linux - starting it from zsh shell - now I'll try bash

**Comment by BurningTreeC** [comment]
bash is just the same

**Comment by BurningTreeC** [comment]
@JohnHind - maybe I'm not the best test-person for it, I've updated my TiddlyDesktop to nw.js v0.30.0 - the current TD is using v0.28.0

**Comment by BurningTreeC** [comment]
@all - can you open another wiki in another process while one is still running?

**Comment by strider72** [comment]
Not working for me. :-(

`"C:\Program Files\TiddlyDesktop\tiddlydesktop-win64-v0.0.13\nw.exe" U:/path/to/mywiki.html`

**Comment by Jermolene** [comment]
Thanks everyone, and apologies for the delay.

**Pull request merged**